File: mobile/src/main/java/com/alexstyl/specialdates/search/ContactEventViewModelFactory.java
package com.alexstyl.specialdates.search;

import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.date.ContactEvent;

import java.util.ArrayList;
import java.util.List;

final class ContactEventViewModelFactory {

    private final EventLabelCreator eventLabelCreator;

    ContactEventViewModelFactory(EventLabelCreator eventLabelCreator) {
        this.eventLabelCreator = eventLabelCreator;
    }

    List<ContactEventViewModel> createViewModelFrom(List<ContactWithEvents> contactEvents) {
        List<ContactEventViewModel> models = new ArrayList<>();
        for (ContactWithEvents contactEvent : contactEvents) {

            Contact contact = contactEvent.getContact();

            for (ContactEvent event : contactEvent.getEvents()) {
                String eventLabel = eventLabelCreator.createFor(event);
                ContactEventViewModel viewModel = new ContactEventViewModel(contact, eventLabel, event.getType().getColorRes(), 5);
                models.add(viewModel);
            }
        }
        return models;
    }
}

File: memento/src/main/java/com/alexstyl/specialdates/search/ContactEventViewModelFactory.kt
package com.alexstyl.specialdates.search

import com.alexstyl.resources.Colors
import com.alexstyl.specialdates.date.ContactEvent

class ContactEventViewModelFactory(private val eventLabelCreator: ContactEventLabelCreator,
                                   private val colors: Colors) {

    fun createViewModelFrom(contactEvents: List<ContactWithEvents>): List<ContactEventViewModel> {
        val models = ArrayList<ContactEventViewModel>()
        for (contactEvent in contactEvents) {

            val contact = contactEvent.contact

            contactEvent.events.forEach { event ->
                val eventLabel = eventLabelCreator.createFor(event)
                val variant = getVariationFor(event)
                models.add(ContactEventViewModel(
                        contact,
                        contact.displayName.toString(),
                        contact.imagePath,
                        eventLabel,
                        colors.getColorFor(event.type),
                        variant
                ))
            }
        }
        return models
    }

    private fun getVariationFor(event: ContactEvent): Int {
        return event.contact.contactID.toInt()
    }
}
